src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -427,10 +427,9 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
             MethodSymbol cacheKey = GetCacheKey(delegateType);
 
             BoundLambda returnInferenceLambda;
-            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType)
+            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType && returnInferenceLambda.Symbol.ReturnType == returnType)
```

Comment:
>&& returnInferenceLambda.Symbol.ReturnType == returnType) [](start = 135, length = 57)

Why do we have to do this explicitly? it looks like _returnInferenceCache uses comparer that pays attention to the return type. I believe this condition was removed around the same time when the comparer was modified, it used to ignore return types before.